document runtime and shim configuration and selection (#9145)
containerd/containerd: runtime/v2/README.md

This document describes the major components of a runtime, how they interact with containerd,
how to use different ones and how to invoke them.

To simplify the interaction, runtime v2 introduced a first class shim API for runtime authors to integrate with containerd.
The shim API is minimal and scoped to the execution lifecycle of a container.

This document is split into the following sections:

* [architecture](#architecture) - the major components, their purposes and relationships
* [usage](#usage) - how to invoke specific runtimes, and how to configure them
* [authoring](#shim-authoring) - how to author a shim

## Architecture

### containerd-runtime communication

containerd expects a runtime to implement several container control features, such as start, stop,
etc. The most common runtime used is [runc](https://github.com/opencontainers/runc).

The high-level flow is as follows:

1. client requests from containerd to create a container
1. containerd lays out the container's filesystem, and creates the necessary config information
1. containerd invokes the runtime, such as `runc` to start the container

> However, containerd does not actually directly invoke the runtime, a process that on Linux normally

Reviewer: ```suggestion
However, containerd does not actually directly invoke the runtime engine to run the container, a process that on Linux normally
```

involves a `fork/exec`. Rather, containerd exposes its network communications port, either

> gRPC or its own leaner [ttRPC](https://github.com/containerd/ttrpc), and expects the runtime

Reviewer: ```suggestion
gRPC or its own leaner [ttRPC](https://github.com/containerd/ttrpc), and expects the runtime shim
```

> to communicate with containerd over that port to receive all commands.

Reviewer: ```suggestion
to communicate with containerd over that port to receive all commands and then route the operations to the runtime engine. In this manner the shim becomes the parent of the container(s).
```

> Since the runtime normally is a simple program that does not run long-term as a daemon, but instead

Reviewer: ```suggestion
Since the runtime engine normally is a simple program that does not run long-term as a daemon, but instead
```

> is invoked as once-and-exit for each command, except for launching the task, most runtimes

Reviewer: ```suggestion
is invoked as once-and-exit for each command, most container runtime architecture models
```

> are implemented in two parts:

Reviewer: ```suggestion
are implemented in four parts:
```

> 1. The shim

Reviewer: ```suggestion
1. The heavy client like kubelet/docker/nerdctl/podman/rancher
2. The container runtime manager, daemon, like containerd/cri-o/dockerd
3. The runtime shim
```

> 1. The runtime

Reviewer: ```suggestion
4. The runtime engine

In some cases various components of the architecture are combined, i.e. shim+engine.
```

#### shim

> The runtime shim is what actually is invoked by containerd. It has minimal options on start beyond

Reviewer: ```suggestion
```

> being provided the communications port for containerd and some bare minimum configuration

Reviewer: ```suggestion
```

> information.

Reviewer: ```suggestion
```

> This shim now handles the communication with containerd over ttrpc, and invokes the actual runtime

Reviewer: ```suggestion
The runtime shim communicates with containerd over ttrpc, and invokes the actual runtime engine
```

via fork/exec.

In theory, it is possible for the shim and the runtime to be the same binary and a single process,

> handling the container in separate threads, but in practice, it is easier to have them be separate

Reviewer: ```suggestion
handling the container in separate threads, but in practice they are usually kept separate to integrate with OCI runc compatible runtime engines.
```

> binaries.

Reviewer: ```suggestion
```

containerd passes options to the shim over the ttrpc connection, including which runtime binary
to invoke. These are the `options` for the [`CreateTaskRequest`](https://github.com/containerd/containerd/blob/main/runtime/v2/README.md#container-level-shim-configuration).

> #### runtime

Reviewer: ```suggestion
#### runtime engine
```

> The runtime itself is what actually starts and stops the container, for example `runc`.

Reviewer: ```suggestion
The runtime engine itself is what actually starts and stops the container, for example [runc](https://github.com/opencontainers/runc).
```

# Runtime v2

Runtime v2 introduces a first class shim API for runtime authors to integrate with containerd.

containerd, the daemon, does not directly launch containers. Instead, it acts as a higher-level manager
or hub for coordinating the activities of containers and content, that lower-level
programs, called "runtime shims" and "runtimes", actually implement to start, stop and manage individual containers

> and container groups called pods.

Author: Ah, makes sense to me. The comments on the first submission by @mikebrow on the document suggested I add that here. Can we have consensus on what should be here?

> and container groups called pods.

Reviewer 2: ```suggestion
and groups of containers (e.g. Kubernetes pods).
```

> and container groups called pods.

Author: Done

> For example, containerd will retrieve container image config and its content as layers, lay it out on disk, set up

Author: Is that worth getting into here, as it is beyond the scope of the runtime (which only is invoked once it is laid out, as far as I understand it)? Or are you suggesting we have a quick reference, e.g.:

```
For example, containerd will retrieve container image config and its content as layers, use the snapshotted to lay it out on disk, set up
```

thus at least providing a reference to use for someone who wants to dig into it?

the container's rootfs and config, and then create a shim instance that uses the runtime engine,
for example `runc`, to actually create/start/stop the container.

This document describes the major components of the v2 runtime shim integration model, how the components interact
with containerd and the v2 shim, and how to use and integrate different v2 shims and runtime engines.

To simplify the interaction, runtime v2 introduced a first class v2 shim API for runtime authors to integrate with containerd,
replacing the v1 shim api.
The shim API is minimal and scoped to the execution lifecycle of a container.

This document is split into the following sections:

* [architecture](#architecture) - the major components, their purposes and relationships
* [usage](#usage) - how to invoke specific runtimes, and how to configure them
* [authoring](#shim-authoring) - how to author a v2 shim

## Architecture

### containerd-runtime communication

containerd expects a runtime to implement several container control features, such as create, start and stop.

> The most common runtime used is [runc](https://github.com/opencontainers/runc).

Author: I am not sure what you are suggesting. That the split between runtime engine and runtime shim is not inherent to containerd, which only cares about, "I have a runtime that speaks by API, I talk to it, I don't care how it goes about doing it (split shim and engine, not split, don't care)"? I get that, and I guess (defer to you, of course) that is technically correct.

In practice, it is split, including components shipped by containerd (the runc shim). Again, in the first review, @mikebrow commented to make the split even more clear. So happy to go with what consensus is, as long as we can get that.

> The most common runtime used is [runc](https://github.com/opencontainers/runc).

Reviewer 2: containerd expects the shim to implement [this interface](https://github.com/containerd/containerd/blob/b30e0163ac36c1a193604e5eca031053d62019c5/api/runtime/task/v2/shim.proto#L33-L51).

That _can_ be implemented by a shim which invokes an OCI runtime (such as runc).  It's just that containerd doesn't "expect" it.  So my suggestion is to rephrase this to avoid the implication that containerd expects that split or the specific OCI runtime spec implementation.

> The most common runtime used is [runc](https://github.com/opencontainers/runc).

Author: Let's let @mikebrow weigh in, as he did an extensive and helpful review (not to mention, describing all of this to me on Slack, so I could understand enough to solve my specific issue, which led to my trying to be helpful to future people by documenting it all).

I think the best approach may be my suggestion above: make it clear here what the border between containerd and "runtime" is, reference in a paragraph that many split shim and engine, and then reference details in a different doc.

> The most common runtime used is [runc](https://github.com/opencontainers/runc).

Reviewer: Windows default path right now merges the shim and runc compatible runtime engine into one binary.. and on linux our default v2 path is the way it's described.  Sam is right language should be coached to indicate flexibility in this regard.  E.g. at the top where describing such detail mention what Sam suggests then you can say something like for the default linux shimv2 yada yada the components are split into..

> The most common runtime used is [runc](https://github.com/opencontainers/runc).

Reviewer 2: I'd still like to see this addressed.

> The most common runtime used is [runc](https://github.com/opencontainers/runc).

Author: I am pretty sure we did. Take a look at the latest version.

> The most common runtime used is [runc](https://github.com/opencontainers/runc).

Reviewer 2: The language here still talks about containerd expecting runc to implement container operations, which isn't correct.

> The most common runtime used is [runc](https://github.com/opencontainers/runc).

Author: I think we should just remove `runc` from there, as it is confusing. Coming in a few mins

> The most common runtime used is [runc](https://github.com/opencontainers/runc).

Author: I tried some restructuring there. Have a look.

The high-level flow is as follows:

1. client requests from containerd to create a container
1. containerd lays out the container's filesystem, and creates the necessary config information
1. containerd invokes the runtime, such as `runc` to create/start/stop the container

However, containerd does not actually directly invoke the runtime engine to create/start/stop the container,
a process that on Linux normally

> involves a `fork/exec`. Rather, containerd exposes its network communications port, either

Author: They don't have to, do they? I had thought it could be network socket or unix-domain socket (which arguably is a local network port)?

> involves a `fork/exec`. Rather, containerd exposes its network communications port, either

Reviewer 2: It's required to be a Unix domain socket or named pipe.  TCP sockets are not supported.

> involves a `fork/exec`. Rather, containerd exposes its network communications port, either

Author: Updated. Have a look.

gRPC or its own leaner [ttRPC](https://github.com/containerd/ttrpc), and expects the runtime
shim to communicate with containerd over that port to receive all commands, and then route the operations
to the runtime engine. In this manner, the shim becomes the parent of the container(s).

> Since the runtime engine normally is a simple program that does not run long-term as a daemon, but instead

Author: This was the exact phraseology suggested by @mikebrow to distinguish between the runtime shim and runtime engine. So I will hold off until we have consensus.

is invoked as once-and-exit for each command, most container runtime architecture models
are implemented in two parts:

1. The heavy client like kubelet/docker/nerdctl/podman/rancher

> 1. The container runtime manager daemon, like containerd/cri-o/dockerd

Author: I am ok either way, but these were @mikebrow 's suggestions. So when you have consensus, I will update it as desired.

1. The runtime shim

> 1. The runtime engine

Author: Mike's recommendation.

In some cases, various components of the architecture are combined into a single executable or process, for example,
shim+engine.

#### runtime shim

The runtime shim is what actually is invoked by containerd. It has minimal options on start beyond
being provided the communications port for containerd and some bare minimum configuration
information.

The runtime shim communicates with containerd over ttrpc, and invokes the actual runtime

> via fork/exec.

Author: Done.

In theory, it is possible for the shim and the runtime to be the same binary and a single process,
handling the container in separate threads, but in practice, they usually are kept separate to integrate

> with distinct OCI `runc`-compatible runtime engines.

Author: Updated.

> with distinct OCI `runc`-compatible runtime engines.

Author: Good comment about firecracker, which gets to part of what you raised at the core of your comments: does container care about the split between shim and engine? Not really, it just cares that a "runtime" (which can be composed of 1 part, 2 parts, or 500 parts) communicates with it over ttRPC. In the case of runc, that is shim and engine. But that is a very common pattern, necessary to understand if you want to understand how it is used most commonly in containerd, so it should be documented.

Let's get consensus, and I can update.

Maybe we want to do like before with the snapshotted? Have a clear description in this doc that containerd only cares about the network comms (ttrpc), then a paragraph or two indicating that the most common one actually splits it and why, and reference a separate doc describing it?

> with distinct OCI `runc`-compatible runtime engines.

Reviewer: yeah it's a naming issue the runtime shim + runtime engine (or VM) either separate or in one binary is a runtime and so is the containerd runtime client and daemon. The term runtime is grossly overloaded. Better to mention the requirements .. the most used broad default linux pattern by which all other oci runtimes are tested.. then a couple other widely used patterns .. windows, kata, firecracker, ...

containerd passes options to the shim over the ttrpc connection, including which runtime binary
to invoke. These are the `options` for the [`CreateTaskRequest`](https://github.com/containerd/containerd/blob/main/runtime/v2/README.md#container-level-shim-configuration).

#### runtime engine

The runtime engine itself is what actually starts and stops the container, for example [runc](https://github.com/opencontainers/runc).

### Relationship

Since each shim instance communicates with containerd as a daemon, while parenting containers via invoking independent runtimes,
it is possible to have one shim for multiple containers and invocations. For example,
you could have one `containerd-shim-runc-v2` communicating with one containerd, and it can
invoke ten distinct containers.

It even is possible to have one shim for multiple containers, each with its own actual runtime,
since, as described above, the runtime binary is passed as one of the options in `CreateTaskRequest`.

Practically, containerd does not work this way. Instead, it invokes a new shim for each container,

> leading to a 1:1 relationship between shim and container.

Author: Take a look now.

In the case where containers are grouped together, like in a pod or sandbox, containerd will

> invoke one shim per pod or sandbox, even though there are multiple individual containers.

Author: Ah, so it is the _shim_ that makes the decision.

Let's see if I got this:

1. containerd receives request to create a container
2. containerd launches a shim, which listens on a socket, reports address, exits
3. containerd tells shim over ttrpc, "launch container A"
4. containerd receive request for another container
5. containerd launches a shim, which reports address and exits. **This might be the same address as before or a new one, totally up to the shim; containerd does not care.**
6. containerd tells shim over ttrpc, "launch container B"

Is that it?

> invoke one shim per pod or sandbox, even though there are multiple individual containers.

Author: Yeah, I just reread your previous comment, and that sounds right. Will try to work it in.

> invoke one shim per pod or sandbox, even though there are multiple individual containers.

Reviewer 2: Step 2 & 5 include the container configuration, step 3 & 6 end up being for controlling the specific container's lifecycle.  That allows the shim to decide whether to have a single socket (and single shim) or multiple sockets (which could be multiple shims or a single shim listening on multiple sockets, though I haven't seen one that does that pattern in practice yet).

> invoke one shim per pod or sandbox, even though there are multiple individual containers.

Author: Updated slightly.

The flow, then, is as follows:

1. containerd receives a request to create a container
1. containerd lays out the container's filesystem, and creates the necessary [container config](https://github.com/opencontainers/image-spec/blob/main/config.md) information
1. containerd invokes the shim, passing it the address to containerd's ttrpc port

> 1. The shim connects to containerd, and waits for commands

Author: Good point. Updating.

> 1. The shim connects to containerd, and waits for commands

Author: Done. Take a look and see if that makes sense.

# Runtime v2

Runtime v2 introduces a first class shim API for runtime authors to integrate with containerd.

containerd, the daemon, does not directly launch containers. Instead, it acts as a higher-level manager
or hub for coordinating the activities of containers and content, that lower-level
programs, called "runtimes", actually implement to start, stop and manage containers,
either individual containers or groups of containers, e.g. Kubernetes pods.

For example, containerd will retrieve container image config and its content as layers, use the snapshotter to lay it out on disk, set up
the container's rootfs and config, and then launch a runtime that will create/start/stop the container.

This document describes the major components of the v2 runtime integration model, how the components interact
with containerd and the v2 runtime, and how to use and integrate different v2 runtimes.

To simplify the interaction, runtime v2 introduced a first class v2 API for runtime authors to integrate with containerd,

> replacing the v1 api.

Reviewer 2: ```suggestion
replacing the v1 API.
```

Author: Good point. Done.

* a single binary for runtime that both listens on the socket and creates/starts/stops the container
* a separate shim binary that listens on the socket, and invokes a separate runtime engine that creates/starts/stops the container

The separate "shim+engine" pattern is used because it makes it easier to integrate distinct runtimes implementing the OCI
runtime spec. The ttRPC protocol can be handle via one runtime shim, while distinct runtime engine implementations can
be used, as long as they implement the OCI runtime spec.

#### Architecture

##### runtime shim

The runtime shim is what actually is invoked by containerd. It has minimal options on start beyond
being provided the communications port for containerd and some configuration information.

The runtime shim listens on the socket for ttRPC commands from containerd, and then invokes a separate actual OCI runtime
engine via `fork`/`exec`.

containerd passes options to the shim over the ttRPC connection, including which runtime binary
to invoke. These are the `options` for the [`CreateTaskRequest`](https://github.com/containerd/containerd/blob/main/runtime/v2/README.md#container-level-shim-configuration).

> ##### runtime engine

Reviewer: ```suggestion
##### runtime engine (OCI runc or compatible) 
```

Author: Is that correct? My understanding (now that I understand it), is that the only part set by containerd is the ttRPC API, i.e. that containerd expects to:

1. Start a process
2. Process exits and returns a socket address
3. That socket has a ttRPC listener available

The split between shim and engine is internal to the runtime implementation. It need not by OCI at all, it could be not split, etc. 

Granted the `runc-v2` shim that ships with containerd _does_ specifically shim to an OCI compatible runtime engine, but that is a specific implementation.

Reviewer: yeah that was my point.. you can have a runc compatible or some other runtime

> The runtime engine itself is what actually starts and stops the container.

Reviewer: ```suggestion
The runtime engine itself is what actually starts and stops the container (typically via libcontainer->sys apis).
```

Author: Added.

For example, in the case of [runc](https://github.com/opencontainers/runc), the containerd project provides the shim
as the executable `containerd-shim-runc-v2`. This is invoked by containerd and starts the ttRPC listener.

The shim then invokes the actual `runc` binary, passing it the container configuration, and the `runc` binary
creates/starts/stops the container.

> 

Reviewer: In some cases runtimes have been combined the shim and OCI runtime engine into one binary.  In other cases a layer or proxy is added to relay OCI container operations into a virtual machine.  Still further some runtimes do not implement OCI containers.

Author: I think we have enough language on that, no?

Reviewer: didn't see it described this way..

Author: Take another look at the latest version. It much more clearly states the expected behaviour of the runtime vis-a-vis containerd and its communications channel, vs what happens internally.

Reviewer: ok for now may add more description at a later time

For example:

```bash
$ ctr --runtime io.containerd.runc.v2 run --rm docker.io/library/alpine:latest alpine
```

Will invoke `containerd-shim-runc-v2`.

You can test this by trying another name:

```bash
$ ctr run --runtime=io.foo.bar.runc2.v2.baz --rm docker.io/library/hello-world:latest hello-world /hello
ctr: failed to start shim: failed to resolve runtime path: runtime "io.foo.bar.runc2.v2.baz" binary not installed "containerd-shim-v2-baz": file does not exist: unknown
```

It received `io.foo.bar.runc2.v2.baz` and looked for `containerd-shim-v2-baz`.

You also can override the default configured runtime for the shim, by passing it the `--runc-binary`
option. For example"

```
ctr --runtime io.containerd.runc.v2 --runc-binary /usr/local/bin/runc-custom run --rm docker.io/library/alpine:latest alpine
```

### Configuring Runtimes

You can configure multiple runtimes in containerd's `config.toml` configuration file, by modifying the
section:

```toml
      [plugins."io.containerd.grpc.v1.cri".containerd.runtimes]
```

See [config.toml man page](../../docs/man/containerd-config.toml.5.md) for more details and an example.

#### Using configured runtimes

> **TODO:** Since we select runtimes by passing both the runtime to invoke (normally a shim) and the binary name (e.g. `runc`) - either via gRPC

Reviewer 2: I think we've now answered this in the other discussion.  The config file is used for CRI, which does not allow the user to specify a shim or options directly.

Author: I don't find it so clear, even after the above, at least from the docs. I imagine someone coming at it fresh will be confused.

> The config file is used for CRI, which does not allow the user to specify a shim or options directly.

So in CRI, I can specify only a name, or nothing at all (i.e. default), and no options?

If so:

1. When can I specify specific runtime path/URI and options? Via gRPC? I assume that `ctr` uses that API?
2. Why is the CRI called in the config file `plugins."io.containerd.grpc.v1.cri".containerd.runtimes`?

Reviewer 2: > So in CRI, I can specify only a name, or nothing at all (i.e. default), and no options?

The CRI proto defines a [`runtime_handler` field](https://github.com/kubernetes/cri-api/blob/de5f1318aede866435308f39cb432618a15f104e/pkg/apis/runtime/v1/api.proto#L476) which is either blank or populated with some name.  If that name maps to a runtime configuration that is defined in the containerd config file, then containerd will use that defined config.

> When can I specify specific runtime path/URI and options? Via gRPC? I assume that ctr uses that API?

Yes, `ctr` and `nerdctl` use the containerd thick client library (which is currently the files at the top-level of this repository) and under the hood that speaks containerd's built-in gRPC API.  `ctr` and `nerdctl` do not speak CRI.

> Why is the CRI called in the config file `plugins."io.containerd.grpc.v1.cri".containerd.runtimes`?

I'm not sure I'm understanding this question.

CRI is a specific gRPC API defined by [this proto definition](https://github.com/kubernetes/cri-api/blob/de5f1318aede866435308f39cb432618a15f104e/pkg/apis/runtime/v1/api.proto).  The CRI plugin in containerd implements the interface defined in that file and serves it over containerd's socket.  The CRI plugin accepts configuration via the config file in sections named `plugins."io.containerd.grpc.v1.cri"`.

Author: OK, got it. Removed that and made the last but a bit clearer. I also squashed the two commits.

Anything else?

* a separate shim binary that listens on the socket, and invokes a separate runtime engine that creates/starts/stops the container

The separate "shim+engine" pattern is used because it makes it easier to integrate distinct runtimes implementing a specific runtime
engine spec, such as the [OCI runtime spec](https://github.com/opencontainers/runtime-spec).
The ttRPC protocol can be handled via one runtime shim, while distinct runtime engine implementations can
be used, as long as they implement the OCI runtime spec.

The most commonly used runtime _engine_ is [runc](https://github.com/opencontainers/runc), which implements the
[OCI runtime spec](https://github.com/opencontainers/runtime-spec). As this is a runtime _engine_, it is not
invoked directly by containerd; instead, it is invoked by a shim, which listens on the socket and invokes the runtime engine.

#### shim+engine Architecture

##### runtime shim

The runtime shim is what actually is invoked by containerd. It has minimal options on start beyond
being provided the communications port for containerd and some configuration information.

> The runtime shim listens on the socket for ttRPC commands from containerd, and then invokes a separate actual OCI runtime
> engine via `fork`/`exec`.

Reviewer 2: This behavior is true of the `io.containerd.runc.v2` shim, but not necessarily of others.

```suggestion
The runtime shim listens on the socket for ttRPC commands from containerd, and then may invoke a separate program to run the container. For example, the `io.containerd.runc.v2` shim invokes the runc OCI runtime
engine via `fork`/`exec`.
```

Author: I took your suggestion and put it in, slightly differently. This is the section of `shim+engine Architecture`, so then it always invokes a separate runtime engine. The "single binary" variant is not covered here (not much to cover).

> containerd passes options to the shim over the ttRPC connection, including which runtime binary
> to invoke. These are the `options` for the [`CreateTaskRequest`](https://github.com/containerd/containerd/blob/main/runtime/v2/README.md#container-level-shim-configuration).

Reviewer 2: ```suggestion
containerd passes options to the shim over the ttRPC connection.  For the `io.containerd.runc.v2` shim these options can include which runtime binary
to invoke. These are the `options` for the [`CreateTaskRequest`](https://github.com/containerd/containerd/blob/main/runtime/v2/README.md#container-level-shim-configuration).
```

Author: OK

##### runtime engine

> The runtime engine itself is what actually starts and stops the container, typically via `libcontainer`->system apis.

Reviewer 2: I'd remove this sentence.  `libcontainer` is only used by runc.

Author: I moved it around. Have a look.
